### ISSUE TITLE
Fastavro fallback doesn't prevent errors

### ISSUE THREAD

**Issue by AlexeyRaga** [confluentinc/confluent-kafka-python#112]
It looks like the intent in `message_serializer.py` is to use `fastavro` and then to fallback to `avro.io` if it is not available.

Because `fastavro` doesn't seem to be able to fully replace `avro.io` and can fail deserialising messaged that are otherwise deserialised fine with `avro.io` the fallback should be more reliable.

**Currently**:
if `fastavro` is present, create a function using it and register as a deserialiser, otherwise register a function that uses `avro.io`

**Ideally**:
Give users an option on whether to use `fastavro` or not by not including it in dependencies in `setup.py` so `pip` does not automatically install one

**Acceptable**:
Register a decode function that tries `fastavro` first, and if it fails fallbacks to `avro.io` for every message.

**Comment by ewencp** [comment]
@AlexeyRaga I get the reasoning, but why would `fastavro` be available at all if it wasn't expected to work? `fastavro` seems like an opt-in optimization? I think it's fine doing something more nuanced if both libraries might be expected to be available and both don't provide the same features. But given the way python is deployed modernly, I would be surprised this would be an issue.

Is this due to virtualenvs where you can't avoid dependency conflicts? Or non-virtualenv environments that include both dependencies? Or something else? I think the additional complexity makes sense if it is unavoidable, but I'd like to understand why `fastavro` becomes a problem before buying into a particular solution to that problem.

**Comment by AlexeyRaga** [comment]
I don't have the dependency conflict. 
When you do `pip install confluent-kafka` it installs `fastavro` with it. Which is not a problem.
But then `message_serializer.py` from this repo does this:

```
HAS_FAST = False
try:
    from fastavro.reader import read_data

    HAS_FAST = True
except:
    pass
```

So it checks if `fastavro` is available (and it always is, because it is in `pip`'s dependencies) it uses it everywhere after because `HAS_AVRO` is `True` now.

And `fastavro` don't seem to be able to handle all the avro cases. And when it can't you have a problem: there is no "runtime" per-message fallback to "normal" avro. It will always try using this "optimisation".

So how would you solve this? 

So short answers:

> why would fastavro be available at all
Because it is in `pip` dependencies with this library.

> if it wasn't expected to work
It is expected, and I'd love it to work, but it doesn't _always_ work

> `fastavro` seems like an opt-in optimization?
I'd hope to. When it doesn't work, how do I opt-out?

**Comment by ewencp** [comment]
@AlexeyRaga I think I see now. So basically you want https://github.com/confluentinc/confluent-kafka-python/blob/master/setup.py#L27 adjusted to remove `fastavro` so you have to explicitly install it to opt in? That seems reasonable and a trivial fix -- let me know if that would resolve your problem (maybe pip uninstall the dependency to verify everything still works ok).

**Comment by edenhill** [comment]
ping @AlexeyRaga , are you happy with @ewencp 's suggested fix?

**Comment by AlexeyRaga** [comment]
@edenhill yes, removing the fastavro dependency seems reasonable to me and being able to opt-in would make it cleaner.
Sorry for the late response, I missed the prev. notification.

**Comment by kkirsanov** [comment]
@AlexeyRaga 
--And fastavro don't seem to be able to handle all the avro cases.
Can you give an example?

For now (fastavro 0.14.9) fastavro fully supports logical types (timestamp-millis/miros, time-millis/micros, date, decomal, UUID)

**Comment by AlexeyRaga** [comment]
@kkirsanov Not really. I only tried it in January 2017 and at the time it couldn't handle some of our messages. I didn't dig into it because all the other Avro serialisers could handle them properly, but not `fastavro`. 
It could be the size of the data as well because some of our payloads are quite big.
It is also possible that since then `fastavro` has improved.

I am not really interested in it anymore since we've moved on to native Haskell implementation for Avro serialisation which seems to be both fast and correct ;)
